Add initial project

hmstepanek opened the pull request:
Add initial simple python server and nginx configuration.

BigBlueHat commented:
@ikreymer you rock. :guitar:

Commit:
Add nginx, dockerfile & python env files

Commit:
Add simple python web server

seanh commented on .coveragerc:
```diff
@@ -0,0 +1,11 @@
+parallel = True
+source =
+    proxy_server
+    tests/proxy_server
```
```suggestion
    tests/unit
```

In lms we've [moved the unit tests into `tests/unit/`](https://github.com/hypothesis/lms/tree/master/tests/unit) (so `tests/unit/proxy_server/...`) so files like `.coveragerc` can just say `tests/unit` which means they're the same for every project

hmstepanek replied:
I noticed in the other repo it looks like the project files moved into a src directory? Would you like me to do that here too?

seanh replied:
@jon-betts Can you remember what we decided about `src` directories? We're using them for the libs but not yet for the apps. I can't remember whether we decided we wanted to move the apps into `src` dirs

jon-betts replied:
I think it makes less sense in the case of an app for sure. The argument for it is just that our scripting and boilerplate might be more similar. Then again, I think the divide between the apps and libs is likely to be large enough that we'll be maintaining two sets regardless.

**Edit to come down on the side of the fence:** I think apps should live in the root. The deployment method here is docker, so things can and do get copied from all over the shop. The reason (apart from random convention) for libraries to live in `src` is because that's *everything* that's going in the library in there. In the case of apps, who knows. It doesn't indicate that and it brings an extra directory in which hints at it.

If we end up using a similar cookiecutter method then that supports updating boilerplate with variables, so a directory which matches the project slug is ok.

seanh replied:
Ok, so @hmstepanek no `src` dir needed

seanh commented on .eslintrc:
```diff
@@ -0,0 +1,3 @@
+{
+  "extends": "hypothesis"
+}
```
Probably don't need eslint yet if we don't have any JS

seanh commented on .prospector.yaml:
```diff
@@ -0,0 +1,43 @@
+output-format: grouped
```
Can we copy the PyLint setup from h-matchers instead of using Prospector? We're trying to move everything over to PyLint but this hasn't been done for the other apps yet.

* `.pylintrc`: https://github.com/hypothesis/h-matchers/blob/master/.pylintrc
* Separate pylintrc for tests: https://github.com/hypothesis/h-matchers/blob/master/tests/.pylintrc
* pydocstyle config file: https://github.com/hypothesis/h-matchers/blob/master/.pydocstyle
* tox installs pylint and pydocstyle: https://github.com/hypothesis/h-matchers/blob/2cca71bef1c6ec85979cf3404000bcdc8a4c84bb/tox.ini#L22-L23
* tox runs pylint and pydocstyle: https://github.com/hypothesis/h-matchers/blob/2cca71bef1c6ec85979cf3404000bcdc8a4c84bb/tox.ini#L40-L43
* No prospector

hmstepanek replied:
I had a query on slack about this [here](https://hypothes-is.slack.com/archives/C1MA4E9B9/p1575595245079900).

seanh replied:
Let's disable D213 as the matches PEP 257

seanh commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+
+# Install nginx & supervisor
+RUN apk add --no-cache nginx gettext supervisor \
+    && mv /usr/bin/envsubst /usr/local/bin/envsubst \
```
?

hmstepanek replied:
It was having trouble finding the command without this and I pulled this from the official nginx image. I didn't bother thinking to deeply about it at the time as I was just trying to get something working first.

hmstepanek replied:
I've removed this and specified an absolute path.

seanh commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+# Install nginx & supervisor
+RUN apk add --no-cache nginx gettext supervisor \
+    && mv /usr/bin/envsubst /usr/local/bin/envsubst \
+    && mkdir -p /run/nginx
```
?

hmstepanek replied:
Apparently this directory doesn't exist on the latest alpine containers so we have to add it manually. Alternatively we can specify an alternative location in the nginx.conf. See [answer here](https://superuser.com/questions/1306656/cant-run-nginx-in-alpine-linux-docker).

hmstepanek replied:
I've moved the path to file into the nginx.conf.

seanh commented on Jenkinsfile:
```diff
@@ -0,0 +1,39 @@
+            sh 'pip install -q tox>=3.8.0'
+            sh 'cd /var/lib/py_proxy && tox -e py36-tests'
+        }
+    }
```
We'll need to run code formatting, lint and coverage on Jenkins as well. See LMS for an example: https://github.com/hypothesis/lms/blob/master/Jenkinsfile

seanh commented on conf/supervisord.conf:
```diff
@@ -0,0 +1,19 @@
+stderr_events_enabled=true
+
+[program:web]
+command=gunicorn -b 0.0.0.0:9082 py_proxy.app:app() -b unix:/tmp/gunicorn-web.sock
```
We've just recently had to change `supervisord.conf` in h and lms to run pserve instead of gunicorn, due to a gunicorn upgrade: https://github.com/hypothesis/lms/blob/e7a72fdf884d924bbc7f7f783e3bea470376a1eb/conf/supervisord.conf#L8

hmstepanek replied:
Do I have to add a prod.ini file for this? It kinda looks that way but I wasn't sure. Anyways-didn't get to this today. I'll see what I can do about this tomorrow.

seanh replied:
Not sure. Depends whether `pserve` requires one, I guess. Actually, now that I think about it, bouncer doesn't have any `.ini` files and as a result it [gets away with calling `gunicorn` directly](https://github.com/hypothesis/bouncer/blob/966117c5fb8e2922b6b14e8094d2a1afc2d6ed6c/conf/supervisord.conf#L8) (since ini file parsing is the reason why `pserve` is needed). So if the new app doesn't need an ini file this should be fine.

seanh commented on py_proxy/__init__.py:
```diff
@@ -0,0 +1 @@
+# -*- coding: utf-8 -*-
```
```suggestion
```

These coding comments aren't need in Python 3

seanh commented on py_proxy/app.py:
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
```
```suggestion
```

seanh commented on py_proxy/views.py:
```diff
@@ -0,0 +1,12 @@
+# -*- coding: utf-8 -*-
```
```suggestion
```

seanh commented on tests/__init__.py:
```diff
@@ -0,0 +1 @@
+# -*- coding: utf-8 -*-
```
```suggestion
```

seanh commented on tests/py_proxy/__init__.py:
```diff
@@ -0,0 +1 @@
+# -*- coding: utf-8 -*-
```
```suggestion
```

seanh commented on tests/py_proxy/conftest.py:
```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
```
```suggestion
```

seanh commented on tests/py_proxy/views_test.py:
```diff
@@ -0,0 +1,9 @@
+# -*- coding: utf-8 -*-
```
```suggestion
```

seanh commented on setup.cfg:
```diff
@@ -0,0 +1,9 @@
+based_on_style = pep8
+
+[flake8]
+max-line-length = 160
```
I don't think this `setup.cfg` file is needed, let's delete it

seanh commented on tox.ini:
```diff
@@ -0,0 +1,58 @@
+    {tests,docstrings,checkdocstrings}: mock
+    lint: flake8
+    {format,checkformatting}: black
+    {format,checkformatting}: isort
```
I think you need an isort config file like this: https://github.com/hypothesis/lms/blob/master/.isort.cfg

seanh commented on tox.ini:
```diff
@@ -0,0 +1,58 @@
+    dev: gunicorn
+commands =
+    dev: {posargs:gunicorn -b "0.0.0.0:9082" --reload "py_proxy.app:app()"}
+    lint: flake8 .
```
This needs to run pylint not flake8 (see comment above with links to how it's done in h-matchers)

seanh reviewed (commented):
Various suggestions but this is looking on the right path so far :+1: 

You might run into a problem when you need subs_filter for HTML rewriting. We don't need to deal with this problem now, we can handle it in a later sprint when we need to actually use subs_filter. But I don't know if there's an easy way to get Nginx with subs_filter when using Alpine. You can get it by using Ubuntu as the base image (`RUN apt-get install -y nginx libnginx-mod-http-subs-filter`) but then you're stuck with whatever version of Python 3 Ubuntu provides, we lose the proper Python versioning that we get from `python:3.6.9-alpine3.10`. I think that's probably fine: this one app can just use Ubuntu's version of Python 3 for now. In the future we'll split it out and not run Python and Nginx in the same Docker container. We'll probably move the Python part into bouncer. Then we won't have the problem anymore. But in the short term Ubuntu is fine.

None of this is a concern right now though, you can just keep using Alpine for now, and we'll deal with this when we come to need rewriting.

seanh committed:
Make tox commands nicer

* `-qe` not `-q`
* `py36-` is unnecessary, e.g. just `tox -qe dev` not `tox -qe py36-dev`
* `-e test` is unnecessary, `test` is the default
* Also prefix all Makefile commands with `@` to make make's output nicer

For more information see https://github.com/hypothesis/h/pull/5866

seanh committed:
Don't ignore the .python-version file

This file is supposed to be committed.

seanh committed:
Add the missing .python-version file

seanh committed:
Simplify .gitignore

There was lots of unnecessary stuff in here. Let's only ignore what we
need to.

seanh committed:
Get test coverage to 100%

These tests don't add much but they get the coverage to 100% which is
useful in and of itself.

seanh committed:
Fix .isort.cfg

seanh committed:
Remove unnecessary comments etc from nginx.conf

seanh commented on py_proxy/views.py:
```diff
@@ -0,0 +1,23 @@
+@view.view_config(renderer="py_proxy:templates/index.html.jinja2", route_name="index")
+def index(request):
+    """Index endpoint."""
+    # pylint: disable=unused-argument
```
You can also rename the argument to `_request`, then you don't need the pylint comment

seanh commented:
I've pushed some commits to this PR to tidy up a few things:

* Nicer tox commands (just added to the rest of our apps today too)
* Added missing `.python-version` file
* Minimal `.gitignore`
* Get test coverage to 100% (this avoids the problem that, for example, a PR that just deletes code can end up reducing test coverage and failing on CI; we'd like all our apps to be at 100% for this reason but the others aren't quite there yet)
* `.isort.cfg` fix
* More minimal `nginx.conf` (just deleted some comments and commented-out things)

seanh reviewed (commented):
I've given this a test locally--all the `Makefile` commands work, I can see the hello world page at http://localhost:9082/ and the health check endpoint at http://localhost:9082/_status. Also after running `make services` I can proxy static files (e.g. <http://localhost:9083/proxy/static/http://example.com>).

I've also taken a look over everything and it looks all good to me :+1:

I think @jon-betts had a couple of small bits of feedback about the nginx.conf, and we need to get Jenkins passing

jon-betts commented on docker-compose.yml:
```diff
@@ -0,0 +1,13 @@
+    volumes:
+      - ./nginx.conf:/etc/nginx/nginx.conf
+      - ./nginx_envsubst.conf.template:/var/lib/hypothesis/nginx_envsubst.conf.template
+    command: bin/sh -c "envsubst '$${ACCESS_CONTROL_ALLOW_ORIGIN}' < /var/lib/hypothesis/nginx_envsubst.conf.template > /var/lib/hypothesis/nginx_envsubst.conf && exec nginx"
```
Is this required or can we removed the command and rely on the `CMD` in the docker file? I tried it locally and it seems to work?

hmstepanek replied:
This is referencing the official nginx image not the custom Dockerfile we created in this repo. (Tested via make services.) I guess if the official image has that exact command but I doubt that’s the case as that template file it’s replacing env vars in is AFAIK a custom name.

jon-betts replied:
Right because this is just used to run this locally, and this avoids having to install docker manually.

jon-betts commented on Dockerfile:
```diff
@@ -0,0 +1,31 @@
+
+USER hypothesis
+
+CMD ["/bin/sh", "-c", "/usr/bin/envsubst '$${ACCESS_CONTROL_ALLOW_ORIGIN}' < /var/lib/hypothesis/nginx_envsubst.conf.template > /var/lib/hypothesis/nginx_envsubst.conf && /usr/bin/supervisord -c /var/lib/hypothesis/conf/supervisord.conf"]
```
This seems to work for me locally, and is simpler. If we are calling shell right away, then we don't lose anything using the shell CMD variant.

```suggestion
CMD /usr/bin/envsubst '$${ACCESS_CONTROL_ALLOW_ORIGIN}' < /var/lib/hypothesis/nginx_envsubst.conf.template > /var/lib/hypothesis/nginx_envsubst.conf && /usr/bin/supervisord -c /var/lib/hypothesis/conf/supervisord.conf
```

hmstepanek replied:
Hmm. I thought when I tried this it didn’t actually run supervisor. I can check it again tho.

hmstepanek replied:
You're right seems to work. Maybe it's cause I was still trying to run with the brackets around the command.

jon-betts commented on Dockerfile:
```diff
@@ -0,0 +1,31 @@
+COPY ./conf/supervisord.conf ./conf/supervisord.conf
+
+COPY ./nginx.conf /etc/nginx/nginx.conf
+COPY ./nginx_envsubst.conf.template /var/lib/hypothesis/nginx_envsubst.conf.template
```
Could these guys get tucked away in `conf/` to keep the root tidy?

It's probably not required as there aren't many files, but a nice pattern I've used before is to replicate the root layout you want in the container in a directly. E.g.

```
conf/
    etc/nginx/
        nginx.conf
    var/lib/hypothesis/
        nginx_envsubst.conf.template
        conf/
           supervisord.conf
```
Then you can do a single copy (permissions permitting) and also review the final locations by looking at the folder.

I also noticed that we copied `/conf/supervisord.conf` relatively and `/nginx_envsubst.conf.template` absolutely although they both end up in `/var/lib/hypothesis/` (I think?)

hmstepanek replied:
Yes putting them in their own area is prob best.

jon-betts commented on Dockerfile:
```diff
@@ -0,0 +1,31 @@
+RUN apk add --no-cache --virtual build-deps \
+  && pip install --no-cache-dir -U pip \
+  && pip install --no-cache-dir -r requirements.txt \
+  && apk del build-deps
```
Nice self contained tidy up here! :+1:

seanh committed:
Fix Jenkinsfile

seanh commented:
Jenkins fixed. @hmstepanek note that I force-pushed this branch to fix the Jenkinsfile

Commit:
fixup: simplifiy docker command

Commit:
fixup: use _ to avoid pylint unused var

Commit:
fixup: move nginx files into nginx directory

jon-betts commented on Dockerfile:
```diff
@@ -0,0 +1,31 @@
+
+COPY ./nginx/nginx.conf /etc/nginx/nginx.conf
+COPY ./nginx/nginx_envsubst.conf.template /var/lib/hypothesis/nginx_envsubst.conf.template
+COPY . .
```
I feel like there is still some tidy up possible here. Do we need to copy everything, or just the python? Might make for a smaller image as well.

hmstepanek replied:
Agh. Idk where that copy . . came from. No we don’t need- I’ll fix that.

seanh replied:
We do need `Makefile`, `tox.ini` and others (linting config files, coverage config files, ...) for Jenkins (it runs the tests, formatting, linting, coverage, etc in the production Docker container).

You may question why these files are in our production Docker image when they're not used in production (only on Jenkins) but [this is how all our apps do it](https://github.com/hypothesis/h/blob/a01ec9315f02659e19672bd0de35d325ca5a90c7/Dockerfile#L61-L62) so I'd recommend fixing it separately and for all the apps together.

Presumably the correct fix is for `Dockerfile` to _not_ copy files that're only needed for CI, and for `Jenkinsfile` to copy those files into the Docker container itself. But we don't need to hold up this PR on getting that working.

hmstepanek replied:
It's definitely wonky but I'm gonna go ahead and merge this as is then.
